### ISSUE TITLE
Add support for MySQL's multi table DELETE statement

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#5230]
MySQL knows statements like:

```sql
DELETE t1, t2 FROM t1 INNER JOIN t2 INNER JOIN t3
WHERE t1.id=t2.id AND t2.id=t3.id;
```

jOOQ should support this syntax.

----
As requested here:
https://groups.google.com/forum/#!topic/jooq-user/EZ81HWhUjkQ

**Comment by rwiggins** [comment]
Note, attempting to use `.join()` with `.delete()` in MySQL produces a SQL syntax exception, e.g.:

    ctx.delete(MY_TABLE.join(MY_OTHER_TABLE).on(...)).where(...).execute();

because it generates SQL like:

    DELETE FROM MyTable JOIN MyOtherTable ON ... WHERE ...;

So, in my opinion this isn't just an Enhancement as tagged, but also a bug — jOOQ that compiles correctly ends up generating invalid MySQL and causing a runtime exception (!).

See the [MySQL manual](http://dev.mysql.com/doc/refman/5.7/en/delete.html) for the correct syntax for multi-table deletes. I'm not sure how the behavior described in the manual can be added to jOOQ cleanly, though, as the MySQL syntax offers a way (1) to delete rows that match from specific table(s) in the join, or (2) to delete rows that match from all tables in the join.

**Comment by lukaseder** [comment]
@rwiggins: We could certainly treat this as two things: An enhancement and that bug you've discovered, although, do note that you can put a `JOIN` expression (which is again an `org.jooq.Table`) in many places where joins are not allowed by some databases, including, for instance, `INSERT`.

At some point, the internal DSL will inevitably become "leaky", because it cannot possibly model all the subtleties of every SQL dialect.

In this case, the effort of improving things is just very high compared to the relatively slim benefit of supporting some obscure vendor-specific syntax...

**Comment by javolek** [comment]
There seems to be already a support for `USING` keyword for `DELETE` statement in JOOQ, since this keyword is used by other dialects too, but the rendered statement in mysql/mariadb dialect is not correct if the table to delete from has an alias. Otherwise it works fine already.

Example:
```
deleteFrom(TABLE_A)
    .using(TABLE_A.join(TABLE_B).on(TABLE_A.ID.eq(TABLE_B.A_ID)))
    .WHERE(TABLE_B.ID.eq(val(123)));
```

**Comment by lukaseder** [comment]
Well, there's a decision to be made if users use `JOIN` in `DELETE FROM` in jOOQ. Did they mean to perform a multi-table `DELETE` or a single table `DELETE`? Hard to say, unless we formally support the multi table `DELETE` syntax (#5230).

But if we support `USING` similar to PostgreSQL (https://github.com/jOOQ/jOOQ/issues/11925), then the intent is clearly to perform a single table `DELETE`, as in @javolek's example.

Perhaps, implementing #11925 should be the first priority here.

**Comment by lukaseder** [comment]
Well, the `UPDATE` statement permits multi table `UPDATE` in the presence of `JOIN`, without the need of explicitly listing the tables to be updated. The difference is that this isn't necessary as the `SET` clause will affect that set of tables implicitly.

For syntactic consistency, I would say that we'll always perform a multi table `DELETE` on all tables involved with a `JOIN` expression by default, until this feature here is implemented.

**Comment by lukaseder** [comment]
@cpech I don't think this is related. This issue here is about multi table `DELETE` statements, where data is deleted from multiple tables. Please create a new feature request for what you're asking
